clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java

```diff
@@ -0,0 +1,124 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.NoSuchElementException;
+
+/**
+ * A class loader that looks for classes and resources in a specified class path first, before delegating to its parent
+ * class loader.
+ */
+public class ChildFirstClassLoader extends URLClassLoader {
+    /**
+     * @param classPath Class path string
+     * @param parent    The parent classloader. If the required class / resource cannot be found in the given classPath,
+     *                  this classloader will be used to find the class / resource.
+     */
+    public ChildFirstClassLoader(String classPath, ClassLoader parent) {
+        super(classpath2URLs(classPath), parent);
+    }
+
+    static private URL[] classpath2URLs(String classPath) {
+        ArrayList<URL> urls = new ArrayList<>();
+        for (String path : classPath.split(";")) {
+            if (path == null || path.trim().isEmpty())
+                continue;
+            File f = new File(path);
+
+            if (path.endsWith("/*")) {
+                try {
+                    File parent = new File(new File(f.getCanonicalPath()).getParent());
+                    if (parent.isDirectory()) {
+                        File[] files = parent.listFiles((dir, name) -> {
+                            String lower = name.toLowerCase();
+                            return lower.endsWith(".jar") || lower.endsWith(".zip");
+                        });
+                        for (File jarFile : files) {
+                            urls.add(jarFile.getCanonicalFile().toURI().toURL());
+                        }
+                    }
+                } catch (IOException e) {
+                }
+            } else if (f.exists()) {
+                try {
+                    urls.add(f.getCanonicalFile().toURI().toURL());
+                } catch (IOException e) {
+                }
+            }
+        }
+        return urls.toArray(new URL[0]);
+    }
+
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        synchronized (getClassLoadingLock(name)) {
+            Class<?> c = findLoadedClass(name);
+
+            if (c == null) {
+                try {
+                    c = findClass(name);
```

Comment:
It's actually very hard to maintain a correct list of system classes / system packages. 

org.apache.kafka.connect.runtime.isolation.PluginUtils has a long list of "system" packages, which contains all Java EE packages, kafka packages, and slf4j packages. Kafka connector plugin class loader does not load classes in those "system packages". I think this is mostly for security reason.

But, Hadoop client library does depend on a different version of java ws.rs / annotation / el / servlet / jsp, and slf4j, which would be blocked by the Kafka connector plugin class loader.

It makes some sense to avoid loading Java SE and Kafka packages.

We can discuss this in another thread, and address the issue in another PR.

Reply:
We should at least avoid java, javax, javafx, jdk, com.sun, org.omg, org.w3c, org.xml, org.ietf packages which are from JDK. Sure, you can address this in another PR.